### ISSUE TITLE
SFT Multi-epoch training

### ISSUE THREAD

**Issue by berserkr** [allenai/open-instruct#646]
Hi, I am seeing LR=0 when doing multi-epoch training. It comes from this line here:

https://github.com/allenai/open-instruct/blob/main/open_instruct/finetune.py#L676C1-L678C6

Here the issue happens when you have multi-node training, so the number of proc > 1, so we need to multiply the max steps by the num of procs as in the else statement regardless. I know the max number of steps is epochs * dataset len / grad acc, however, as it is described in the text, dataset is seen 1/n proc times. However, the lr scheduler is called n proc times per step, hence after a single epoch , LR goes to 0.

So the fix would be to just multiply n procs * max num steps. 

Has anyone else seen LR go to 0 in such multi-epoch, multi-node scenarios?

**Comment by vwxyzjn** [comment]
Need to read over this a bit more. I always find the scheduler logic a bit difficult to predict its behavior.

> Has anyone else seen LR go to 0 in such multi-epoch, multi-node scenarios?

We have quite a few tracked runs that suggest this is not an issue. They are 2 epochs, multi-node settings. 


https://allenai.github.io/open-instruct/algorithms/finetune/#reproduce-allenaillama-31-tulu-3-8b-sft-8-nodes


<img width="1107" alt="Image" src="https://github.com/user-attachments/assets/b9297029-b308-402b-88d2-b3101c6a334f" />

<img width="831" alt="Image" src="https://github.com/user-attachments/assets/655b6a5c-2f0e-41d6-baf2-0817ff36445f" />

**Comment by finbarrtimbers** [comment]
I am going to close this as we have a lot of successful multi-node training runs. If you keep encountering this please let me know and we can look into this more.
